lib/email_messages.py

```diff
@@ -61,3 +62,28 @@ def email_course_error(request, exercise, message, exception=True):
         email_course_instance(instance, subject, body)
     except:
         pass
+
+
+def email_course_students(
+        instance: CourseInstance,
+        subject: str,
+        message: str,
+        audience: CourseInstance.ENROLLMENT_AUDIENCE=CourseInstance.ENROLLMENT_AUDIENCE.ALL_USERS
+        ) -> bool:
+    """
+    Sends an email to students on the course. Audience parameter controls whether the mail goes
+    to all (default), just internal, or just external students.
+    """
+    students = instance.students
+    if audience == instance.ENROLLMENT_AUDIENCE.INTERNAL_USERS:
+        students = students.filter(organization=settings.LOCAL_ORGANIZATION)
+    if audience == instance.ENROLLMENT_AUDIENCE.EXTERNAL_USERS:
+        students = students.exclude(organization=settings.LOCAL_ORGANIZATION)
+
+    recipients = students.exclude(user__email='').values_list("user__email", flat=True)
+
+    try:
+        return send_mail(subject, message, settings.SERVER_EMAIL, recipients, True) == 1
```

Comment:
Did you notice that Django has a function `send_mass_mail` that is specifically intended for mass mailing, e.g., 1000 students at a time? `send_mass_mail` is slightly more efficient.

With `send_mass_mail`, we can also easily avoid including all students in the "To" header, which would be visible to all students in the course (leaking all student email addresses to all students).

Note that `send_mass_mail` may return larger integers than just `1`.

https://docs.djangoproject.com/en/3.2/topics/email/#send-mass-mail

--------------------

`recipients` 
* in the docs, `recipient_list` parameter: A list of strings, each an email address. Each member of `recipient_list` will see the other recipients in the “To:” field of the email message.
* If there are 1000 students, I assume it is problematic to show 1000 recipients in the "To" field in any email client.
* At any rate, students' emails must not be revealed to all other students in the course. It is personal data and we can't leak it to everyone like this. It is not useful to list the emails to everyone like this. **This is a privacy breach.** Imagine if someone "replies to all" in the email: not very nice.

`fail_silently`
* Did you intend to use `fail_silently=False` (the default)?
* Why would we hide the raised exception? The exception can be used to log more details about the error. It will help the server admins.
* You already have the `except` block, but you specifically silenced the exceptions that could be raised.
* Let's use `fail_silently=False`, which is the default.

https://docs.djangoproject.com/en/3.2/topics/email/#send-mail